src/conversations/Conversation.ts

```diff
@@ -46,7 +67,177 @@ export default class Conversation {
+   * Send a message into the conversation
+   */
+  async send(
+    message: any, // eslint-disable-line @typescript-eslint/no-explicit-any
+    options?: SendOptions
+  ): Promise<MessageV2> {
+    const msg = await this.encodeMessage(message, options)
+    this.client.publishEnvelopes([
+      {
+        contentTopic: this.topic,
+        message: msg.toBytes(),
+        timestamp: msg.sent,
+      },
+    ])
+    return msg
+  }
+
+  private async encodeMessage(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    content: any,
+    options?: SendOptions
+  ): Promise<MessageV2> {
+    const payload = await this.client.encodeContent(content, options)
+    const header: xmtpEnvelope.MessageHeaderV2 = {
+      topic: this.topic,
+      createdNs: dateToNs(options?.timestamp || new Date()),
+    }
+    const headerBytes = xmtpEnvelope.MessageHeaderV2.encode(header).finish()
+    const digest = await sha256(concat(headerBytes, payload))
+    const signed = {
+      payload,
+      sender: this.client.keys.getPublicKeyBundle(),
+      signature: await this.client.keys.getCurrentPreKey().sign(digest),
+    }
+    const signedBytes = xmtpEnvelope.SignedContent.encode(signed).finish()
+    const ciphertext = await encrypt(signedBytes, this.keyMaterial, headerBytes)
+    const protoMsg = {
+      v1: undefined,
+      v2: { headerBytes, ciphertext },
+    }
+    const bytes = xmtpEnvelope.Message.encode(protoMsg).finish()
+    const msg = await MessageV2.create(protoMsg, header, signed, bytes)
+    return msg
+  }
+
+  private async decodeMessage(env: messageApi.Envelope): Promise<MessageV2> {
+    if (!env.message) {
+      throw new Error('empty envelope')
+    }
+    const msg = xmtpEnvelope.Message.decode(b64Decode(env.message.toString()))
+    if (!msg.v2) {
+      throw new Error('unknown message version')
+    }
+    const msgv2 = msg.v2
+    const header = xmtpEnvelope.MessageHeaderV2.decode(msgv2.headerBytes)
+    if (header.topic !== this.topic) {
+      throw new Error('topic mismatch')
+    }
+    if (!msgv2.ciphertext) {
+      throw new Error('missing ciphertext')
+    }
+    const decrypted = await decrypt(
+      new Ciphertext(msgv2.ciphertext),
+      this.keyMaterial,
+      msgv2.headerBytes
+    )
+    const signed = xmtpEnvelope.SignedContent.decode(decrypted)
+    if (
+      !signed.sender?.identityKey ||
+      !signed.sender?.preKey ||
+      !signed.signature
+    ) {
+      throw new Error('incomplete signed content')
+    }
+
+    const digest = await sha256(concat(msgv2.headerBytes, signed.payload))
+    if (
+      !new SignedPublicKey(signed.sender?.preKey).verify(
+        new Signature(signed.signature),
+        digest
+      )
+    ) {
+      throw new Error('invalid signature')
+    }
+    const message = await MessageV2.create(msg, header, signed, env.message)
+    await this.client.decodeContent(message)
+    return message
   }
 }
+
+export type Conversation = ConversationV1 | ConversationV2
```

Comment:
I looked at the output type of this, and it's exactly what I would have written as an interface. Think it's fine to leave as-is